### PR TITLE
Guard against DoS in name constraints handling.

### PR DESCRIPTION
**Pull request by dot-asm**
This guards against the name constraints check consuming large amounts
of CPU time when certificates in the presented chain contain an
excessive number of names (specifically subject email names or subject
alternative DNS names) and/or name constraints.

Name constraints checking compares the names presented in a certificate
against the name constraints included in a certificate higher up in the
chain using two nested for loops.

Move the name constraints check so that it happens after signature
verification so peers cannot exploit this using a chain with invalid
signatures. Also impose a hard limit on the number of name constraints
check loop iterations to further mitigate the issue.

Thanks to NCC for finding this issue. Fix written by Martin Kreichgauer.

Reviewed-by: Rich Salz <rsalz@openssl.org>
Reviewed-by: Andy Polyakov <appro@openssl.org>
(Merged from https://github.com/openssl/openssl/pull/4393)

(cherry picked from commit 8545051c3652bce7bb962afcb6879c4a6288bc67)

Resolved conflicts:
	crypto/x509v3/v3_ncons.c
	test/recipes/25-test_verify.t

### CODE REVIEW

**Comment by dot-asm** [comment]
It was suggested that this should be applied to 1.1.0, but question also is if 1.0.2 deserves it.

**Comment by dot-asm** [comment]
Oh! Commit message will have to be adjusted to reflect this merge request...

**Comment by richsalz** [comment]
I think the NC code in 1.0.2 was basically broken, but if you can cherry-pick I approve that too

**Comment by dot-asm** [comment]
> if you can cherry-pick I approve that too

It wasn't a coincidence that I wrote "deserve" and not "cherry-pick". This code definitely won't cherry-pick into 1.0.2, but it doesn't answer the question: if this protection is worth considering even in 1.0.2 context. I.e. *irregardless* cherry-pick-ability.

**Comment by richsalz** [comment]
I suppose it should be fixed in 1.0.2

**Comment by dot-asm** [comment]
And this one still needs approval :-)

File: test/recipes/25-test_verify.t [hunk 1]
```diff
@@ -328,3 +328,17 @@ ok(!verify("badalt9-cert", "sslserver", ["root-cert"], ["ncca1-cert", "ncca3-cer
...
+    "Too many names and constraints to check (2)");
+ok(!verify("many-names3", "sslserver", ["many-constraints"], ["many-constraints"], ),
+    "Too many names and constraints to check (3)");
+
+ok(verify("some-names1", "sslserver", ["many-constraints"], ["many-constraints"], ),
+    "Not too many names and constraints to check (1)");
+ok(verify("some-names2", "sslserver", ["many-constraints"], ["many-constraints"], ),
+    "Not too many names and constraints to check (2)");
+ok(verify("some-names2", "sslserver", ["many-constraints"], ["many-constraints"], ),
+    "Not too many names and constraints to check (3)");
```
Comment (levitte) [inline test/recipes/25-test_verify.t]: On all of these, the second line seems misaligned by one space.

Comment (dot-asm) [reply test/recipes/25-test_verify.t]: It's a cherry-pick artefact. I mean one can observe same pattern in the original. And one can note that the pattern apparently was copy-pasted from previous line, i.e. it wasn't introduced here... So it's not sufficient to fix it here...

**Comment by dot-asm** [comment]
Reformat pushed...

**Comment by dot-asm** [comment]
Ping @levitte.

**Pull request closed without merging**